Office365 - Remove restriction of 1000 contacts in search

nissou91 opened the pull request:
**Problem Statement:**
* In Microsoft API, when a user `Alice` has more than 1000 contacts and she tries to get/list all of them using Wazo application; she does not see them all

**Cause/Bottleneck:**
* The issue comes for Office365 API which uses paging [O365-paging](https://learn.microsoft.com/en-us/graph/paging?view=graph-rest-1.0).

**Solution:**
* We have to change the code of [fetch-method](https://github.com/wazo-platform/wazo-dird/blob/master/wazo_dird/plugins/office365_backend/services.py#L35) in `wazo-dird` to support Microsoft API Paging and parse all pages within the returned result.

bloom1 reviewed (changes requested):
I tested the PR for real on `wazo-test`. I added some debug lines to see what was happening and here is the output: 
```
2022-11-28 16:21:56,417 [5995] (ERROR) (wazo_dird.plugins.office365_backend.services): An error occured while fetching information from microsoft endpoint
2022-11-28 16:21:56,417 [5995] (DEBUG) (wazo_dird.plugins.office365_backend.services): Status code: 400, response: "{"error":{"code":"BadRequest","message":"Query option '$top' was specified more than once, but it must be specified at most once.","innerError":{"date":"2022-11-28T21:21:56","request-id":"8de4869f-30c9-45b7-8441-xxxefb82exxx","client-request-id":"8de4869f-30c9-45b7-8441-xxxefb82exxx"}}}"
```

nissou91 commented on integration_tests/suite/helpers/fixtures/http.py:
```diff
@@ -197,6 +197,51 @@ def wrapper(self, *args, **kwargs):
+            count_expectation['times']['unlimited'] = True
+            mock_server.mock_any_response(count_expectation)
+            counter = 0
+            while counter < len(pages):
```
yes you are right; I was using it for another purpose (testing) and forgot to remove the counter; fixed it :)

nissou91 commented on integration_tests/suite/test_office365_contacts.py:
```diff
@@ -79,6 +143,31 @@ def test_unknown_source(self):
+            ),
+        )
+        result_items = result.get('items')
+        for i in range(5000):
```
it is the total number of contacts that I generate (all three pages); I will create a constant for that.

nissou91 commented on integration_tests/suite/test_office365_contacts.py:
```diff
@@ -1,7 +1,14 @@
 # Copyright 2020-2021 The Wazo Authors  (see the AUTHORS file)
```
Thanks, I did not know about this tool; I changed them manually for now :)

nissou91 commented on integration_tests/suite/helpers/fixtures/http.py:
```diff
@@ -197,6 +197,51 @@ def wrapper(self, *args, **kwargs):
     return decorator
 
 
+def office365_paginated_result(pages):
```
Done

nissou91 commented on wazo_dird/plugins/office365_backend/services.py:
```diff
@@ -35,18 +35,69 @@ def get_contacts(self, microsoft_token, url, **list_params):
-                )
-                raise UnexpectedEndpointException(
-                    endpoint=url, error_code=response.status_code
+            if count == 1:
```
Done

nissou91 commented on wazo_dird/plugins/office365_backend/services.py:
```diff
@@ -35,18 +35,69 @@ def get_contacts(self, microsoft_token, url, **list_params):
+                        )
+                        if next_page_response.status_code == 200:
+                            logger.debug(
+                                f'Successfully fetched contacts from microsoft page: {next_page_url}'
```
corrected

bloom1 replied:
Sorry if it was not clear: the formatting should not be applied yet, so the parameter should be `logger.debug('... %s', variable)`

nissou91 replied:
ahh okay, got it

nissou91 replied:
Done

nissou91 commented:
> I tested the PR for real on `wazo-test`. I added some debug lines to see what was happening and here is the output:
> 
> ```
> 2022-11-28 16:21:56,417 [5995] (ERROR) (wazo_dird.plugins.office365_backend.services): An error occured while fetching information from microsoft endpoint
> 2022-11-28 16:21:56,417 [5995] (DEBUG) (wazo_dird.plugins.office365_backend.services): Status code: 400, response: "{"error":{"code":"BadRequest","message":"Query option '$top' was specified more than once, but it must be specified at most once.","innerError":{"date":"2022-11-28T21:21:56","request-id":"8de4869f-30c9-45b7-8441-xxxefb82exxx","client-request-id":"8de4869f-30c9-45b7-8441-xxxefb82exxx"}}}"
> ```

I have removed the `$top` query param from the request and the expectation to avoid redundancy error like the one you got.

nissou91 commented:
recheck

bloom1 reviewed (changes requested):
I think you may want to add more test cases that cover more scenarios.

I am getting the following error:
```
2022-12-02 15:09:39,404 [32472] (ERROR) (root): An error occured in search                                                                                                                                         
Traceback (most recent call last):                                                                                                                                                                                 
  File "/usr/src/wazo/wazo-dird-WP636/wazo_dird/helpers.py", line 46, in execute                                                                                                                                   
    return function(*args, **kwargs)                                                                                                                                                                               
  File "/usr/src/wazo/wazo-dird-WP636/wazo_dird/plugins/office365_backend/plugin.py", line 91, in search                                                                                                           
    sorted_contacts = sorted(filtered_contacts, key=itemgetter('givenName'))                                                                                                                                       
TypeError: '<' not supported between instances of 'str' and 'NoneType'                                                                                                                                             
2022-12-02 15:09:39,419 [32472] (INFO) (flask.app): response to 10.0.4.2 in 0.83s: GET https://wazo-test/api/dird/0.1/directories/lookup/default?term=a 200   
```
Another question: did you test the PR with a real stack and a real office365 account?

nissou91 commented:
> I think you may want to add more test cases that cover more scenarios.
> 
> I am getting the following error:
> 
> ```
> 2022-12-02 15:09:39,404 [32472] (ERROR) (root): An error occured in search                                                                                                                                         
> Traceback (most recent call last):                                                                                                                                                                                 
>   File "/usr/src/wazo/wazo-dird-WP636/wazo_dird/helpers.py", line 46, in execute                                                                                                                                   
>     return function(*args, **kwargs)                                                                                                                                                                               
>   File "/usr/src/wazo/wazo-dird-WP636/wazo_dird/plugins/office365_backend/plugin.py", line 91, in search                                                                                                           
>     sorted_contacts = sorted(filtered_contacts, key=itemgetter('givenName'))                                                                                                                                       
> TypeError: '<' not supported between instances of 'str' and 'NoneType'                                                                                                                                             
> 2022-12-02 15:09:39,419 [32472] (INFO) (flask.app): response to 10.0.4.2 in 0.83s: GET https://wazo-test/api/dird/0.1/directories/lookup/default?term=a 200   
> ```
> 
> Another question: did you test the PR with a real stack and a real office365 account?

I will take a look. and yes, I tried the code on my real Office365 account; where I imported the randomly generated contacts into my account and then run the code. However, I did not run it on a wazo stack...

nissou91 commented:
recheck
